src/engine/SpatialJoinAlgorithms.cpp

```diff
@@ -97,48 +103,43 @@ Point SpatialJoinAlgorithms::convertGeoPointToPoint(GeoPoint point) const {
 // ____________________________________________________________________________
 Id SpatialJoinAlgorithms::computeDist(const rtreeEntry& geo1,
                                       const rtreeEntry& geo2) const {
-  auto convertPoint = [&](AnyGeometry geometry, std::optional<Box> bbox) {
+  auto convertPoint = [&](const AnyGeometry& geometry,
+                          const std::optional<Box>& bbox) {
     Box areaBox;
-    if (bbox.has_value()) {
-      areaBox = bbox.value();
-    } else {
-      areaBox = boost::apply_visitor(BoundingBoxVisitor(), geometry);
-    }
+    areaBox =
+        bbox.value_or(boost::apply_visitor(BoundingBoxVisitor(), geometry));
     Point p = calculateMidpointOfBox(areaBox);
     return GeoPoint(p.get<1>(), p.get<0>());
   };
   // use the already parsed geometries to calculate the distance
   if (useMidpointForAreas_) {
-    auto point1 = geo1.geoPoint;
-    auto point2 = geo2.geoPoint;
+    auto point1 = geo1.geoPoint_;
+    auto point2 = geo2.geoPoint_;
     if (!point1) {
-      point1 = convertPoint(geo1.geometry.value(), geo1.boundingBox);
+      point1 = convertPoint(geo1.geometry_.value(), geo1.boundingBox_);
     }
     if (!point2) {
-      point2 = convertPoint(geo2.geometry.value(), geo2.boundingBox);
-    }
-    if (!point1.has_value() || !point2.has_value()) {
-      return Id::makeUndefined();
+      point2 = convertPoint(geo2.geometry_.value(), geo2.boundingBox_);
     }
     return Id::makeFromDouble(
         ad_utility::detail::wktDistImpl(point1.value(), point2.value()));
   } else {
-    if (geo1.geoPoint.has_value() && geo2.geoPoint.has_value()) {
+    if (geo1.geoPoint_.has_value() && geo2.geoPoint_.has_value()) {
       return Id::makeFromDouble(ad_utility::detail::wktDistImpl(
-          geo1.geoPoint.value(), geo2.geoPoint.value()));
-    } else if (geo1.geometry.has_value() && geo2.geometry.has_value()) {
+          geo1.geoPoint_.value(), geo2.geoPoint_.value()));
+    } else if (geo1.geometry_.has_value() && geo2.geometry_.has_value()) {
       return Id::makeFromDouble(
-          convertDegreesToMeters(geo1.geometry.value(), geo2.geometry.value()));
+          computeDist(geo1.geometry_.value(), geo2.geometry_.value()));
     } else {
       // one point and one area
-      AnyGeometry geom1 = geo1.geometry.value();
-      AnyGeometry geom2 = geo2.geometry.value();
-      if (geo1.geoPoint.has_value()) {
-        geom1 = convertGeoPointToPoint(geo1.geoPoint.value());
-      } else if (geo2.geoPoint.has_value()) {
-        geom2 = convertGeoPointToPoint(geo2.geoPoint.value());
+      std::optional<AnyGeometry> geom1 = geo1.geometry_;
+      std::optional<AnyGeometry> geom2 = geo2.geometry_;
```

Comment:
This copies the geometry. I think you can do `const std::optional<AnyG>& geom1 = ...` or even `const auto& geom1 = ...`

Reply:
I now see what you are doing, and what you are trying to do, and my thing doesn't work.

what about
```  
if (! geo1.geometry_.has_value() { geo1.geometry_ = geoPointToPoint(geo1.geoPoint_.value());
```
(same for `geo2`, then you alsways have the `geometry_` value. 
If you do this directly when reading those geometries from the IdTable, then the `geometry_` member doesn't even have to be `optional`, and this code gets much simpler. Otherwise, if you do it here (if I am missing something obvious), then factor that code snippet into a separate function.